### ISSUE TITLE
simplify repositories and make builds more reliable

### ISSUE THREAD

**Issue by luite** [ghcjs/ghcjs#467]
We should simplify the repository structure as much as possible and use the opportunity to make maintenance easier and reduce the room for mistakes.

main changes:

 * pull the `ghcjs-boot` repository into `ghcjs`, have all upstream libraries as submodules
 * host a mirror of all submodules, rather than referring to arbitrary upstream infrastructure
 * mirrors should never lose commit ids referred to by any ghcjs revision even if upstream rebases or force-pushes (collect submodule hashes from ghcjs repo and auto-push branch heads or tags to keep them alive?)

maintenance improvements:

  * have an easy way to switch to a new upstream commit, apply our patches and use the result as our boot lib. it should remain clear what changes are local to the ghcjs mirror. (this should be used to follow for example the upstream commits from ghc 8.0.1 -> ghc 8.0.2 and its boot libraries)

question:

should we really switch away from patches in the ghcjs-boot repo? the patches are rather ugly, but it does let us keep the same submodule ids as upstream, and it makes the separation between upstream and ours very clear. if we move it into git, then we need some consistent workflow to make sure we don't lose this.

**Comment by wuzzeb** [comment]
One benefit to removing patches and instead using git is that you can keep the changes together and move the javascript and haskell code together.  For example, the `process` fork package has a `pkg/process.js` file in `ghcjs-shims` and then a patch in `ghcjs-boot/patches/process.patch`.  These are wildly separated from each other and make any sort of change difficult since you need to update two files in separate repositories.  If you had a git fork of the process package, you could add the `process.js` file to the `jsbits` of the `process` cabal file and therefore keep the code related to the process package contained.

I think we could even then eliminate `ghcjs-shims`, making the build more reliable and resilient to change.  In fact, just glancing through `ghcjs-shims`, most of the code looks like it could be moved into `jsbits` in various forked packages, with perhaps the bulk of the libs moving into `jsbits` inside the `ghcjs-base` repository.  The only thing left in the `ghcjs-shims` repo is then the RTS itself, but moving the RTS into the base `ghcjs` repository makes a lot of sense for versioning, since the RTS will be tied to the ghcjs version.  

A final benifit to using git over patches is that the eventual goal should be to move as much of this code upstream as possible.  It depends on each upstream maintainer of course if they want to accept it, but if instead of the current situation with files spread out over multiple repositories, you had a git fork of the upstream project that added the `ifdef GHCJS` stuff, the code would be in much better position to send a pull request to the upstream maintainer.  Of course, at the moment probably don't want to push any code upstream until things like `ghcjs-base` stabilize, but if ghcjs is using git forks of each upstream project you are ready when that happy day arrives.

**Comment by luite** [comment]
I agree. Shims is not used for its originally intended purpose, which was having JS support in packages before Cabal support was available. Everything under `/rts` should probably eventually be moved into the `ghcjs` repo, and the `/pkg` things imported into the various upstream packages.

But perhaps it's best to do this as a two-step thing. First pull `ghcjs-boot` in, making the invidual libraries submodules of `ghcjs` (including `shims` perhaps). Then once that works, integrate the jsbits into our forks of the upstream packages, move `/rts` from `shims` into `ghcjs` and eventually decommission `shims`.

But for pre-release testing or making a package compatible with GHCJS after they've been uploaded to hackage I think we don't have a good alternative yet. With `shims`, you just add a yaml file, and all matched versions magically get the required jsbits linked. It's not exactly pretty, but it gets the job done. Therefore I think we  should do `ghcjs-boot` first.

**Comment by Ericson2314** [comment]
All of this sound good! One addendum is that once we are using a package from its hackage release, we can get the code from there so no need to keep around the fork [assuming we use hackage release for the dependency in question before a GHCJS release we want to support long term].

**Comment by Ericson2314** [comment]
CC @cstrahan In case you are interested.

**Comment by Ericson2314** [comment]
Ok I've forked all the repositories which are patched https://github.com/forked-upstream-packages-for-ghcjs (including GHC itself). Next steps are (in rough order):

1. Get rid of ghcjs-boot
 - [ ] Patch forks
 - [ ] Update ghcjs-boot submodules to use forks
 - [ ] Update ghcjs-boot to use GHC fork instead for libraries copied in ghcjs-boot
 - [ ] Investigate shallow-cloning submodules for ghcjs-boot
 - [ ] Investigate Hackage releases for unpatched libraries
 - [ ] Merge ghcjs-boot and ghcjs (`git subtree` is my favorite thing for this)
 - [ ] Use `impl(ghcjs)` instead of `#ifdef`-ing out whole modules
 - [ ] Upstream PRs round 1?

2. Get rid of ghcjs-shims
 - [ ] Add jsbits from shims
 - [ ] Add `/rts` to ghcjs
 - [ ] Upstream PRs round 2?

**Comment by Ericson2314** [comment]
> But for pre-release testing or making a package compatible with GHCJS after they've been uploaded to hackage I think we don't have a good alternative yet. With shims, you just add a yaml file, and all matched versions magically get the required jsbits linked. It's not exactly pretty, but it gets the job done. Therefore I think we should do ghcjs-boot first.

I agree with doing with boot before shims, but I think the solution is edit to edit the cabal file (post release) to signify GHCJS incompatibility, and then make a patch release (or whatever the least-significant section of the release code is called) with the jsbits added.
